### DISCUSSION TITLE
west flash --runner=jlink failes with runner.jlink not found

### DISCUSSION THREAD

**Discussion by emmdaniel** [zephyrproject-rtos/zephyr#66894] [Q&A] [answered]
Hi All,
This is more of a question as maybe I'm doing something wrong.
I setup my zephyr environment yesterday on an Ubuntu 22.04, following the provided user guide, all worked fine, I was able to compile and flash an ESP32 and RPI4B.
I setup also the JLink as I have a dongle around and saw it can be used, so I downloaded the JLinkExe from Segger and installed it.
Tried to see if the connection with the dongle is fine and although I didn't have a target connected to Jlink I could see all went well as it was failing when trying to detect the target. 
Today I got my hands on a RPI Pico and tried to flash the blinky example.
Compilation went fine, but, when trying to flash I get the following error:

```
╰$ west flash --runner jlink                                                                                                                                                          
-- west flash: rebuilding
[1/11] Performing build step for 'second_stage_bootloader'
ninja: no work to do.
[2/2] cd /home/mdaniel/zephyrproject/zephyr/build/zephyr && /home/mdaniel/zephyrproject/.venv/bin/pyt.../build/check_init_priorities.py --elf-file=/home/mdaniel/zephyrproject/zephyr/build/zephyr/zephyr.elf
-- west flash: using runner jlink
-- runners.jlink: JLink version: 7.94b
-- runners.jlink: Flashing file: /home/mdaniel/zephyrproject/zephyr/build/zephyr/zephyr.hex
FATAL ERROR: command exited with status 1: /opt/SEGGER/JLink_V794b/JLinkExe -nogui 1 -if swd -speed auto -device RP2040_M0_0 -CommanderScript /tmp/tmprktku0xojlink/runner.jlink -nogui 1

```
So I ran manually that command and saw it fails because it can't find the `/tmp/tmprktku0xojlink/runner.jlink`
The `/tmp `is not restricted to non **sudo** users, so it's not an access issue. I see the folder name changes with every `west` flash command, so I'm assuming the `runner.jlink` is generated every time. But for some reason it fails and no other error is generated.
I reinstalled the entire setup once again, not helping.

Any suggestion much appreciated.
Thanks.

**Comment by DerekSnell** [comment]
Hi @emmdaniel ,
The JLink commander script is generated, and is removed after the `west flash` command, so it is expected that commander script will be missing afterwards.  When I have issues with the JLink runner, I run the West command in verbose mode by adding the `-v` argument to the command below.  That should display all the output from the JLink commander tool directly, and help identify the issue.
> west -v flash --runner jlink

Hope that helps

**Comment by emmdaniel** [comment]
Thanks @DerekSnell , I didn't write in my original post that I tried already the verbose mode:

```
╰$ west -v flash --runner jlink
-- west flash: rebuilding
cmake version 3.22.1 is OK; minimum version is 3.13.1
Running CMake: /usr/bin/cmake --build /home/mdaniel/zephyrproject/zephyr/build
[1/11] Performing build step for 'second_stage_bootloader'
ninja: no work to do.
[2/2] cd /home/mdaniel/zephyrproject/zephyr/build/zephyr && /home/mdaniel/zephyrproject/.venv/bin/pyt.../build/check_init_priorities.py --elf-file=/home/mdaniel/zephyrproject/zephyr/build/zephyr/zephyr.elf
-- west flash: using runner jlink
runners.jlink: JLINKARM_GetDLLVersion()=79402
-- runners.jlink: JLink version: 7.94b
runners.jlink: JLink commander script:
ExitOnError 1
r
loadfile "/home/mdaniel/zephyrproject/zephyr/build/zephyr/zephyr.hex"
g
writeDP 1 0
readDP 1
q
-- runners.jlink: Flashing file: /home/mdaniel/zephyrproject/zephyr/build/zephyr/zephyr.hex
runners.jlink: /opt/SEGGER/JLink_V794b/JLinkExe -nogui 1 -if swd -speed auto -device RP2040_M0_0 -CommanderScript /tmp/tmpipvbivnyjlink/runner.jlink -nogui 1
SEGGER J-Link Commander V7.94b (Compiled Dec 13 2023 17:05:47)
DLL version V7.94b, compiled Dec 13 2023 17:05:24


J-Link Command File read successfully.
Processing script file...
J-Link>ExitOnError 1
J-Link Commander will now exit on Error
J-Link>r
J-Link connection not established yet but required for command.
Connecting to J-Link via USB...O.K.
Firmware: J-Link ARM V8 compiled Nov 28 2014 13:44:46
Hardware version: V8.00
J-Link uptime (since boot): N/A (Not supported by this model)
S/N: 58000399
License(s): FlashBP, FlashDL, GDB
VTref=0.000V
Target connection not established yet but required for command.
Device "RP2040_M0_0" selected.


Connecting to target via SWD
Cannot connect to target.

Script processing completed.

FATAL ERROR: command exited with status 1: /opt/SEGGER/JLink_V794b/JLinkExe -nogui 1 -if swd -speed auto -device RP2040_M0_0 -CommanderScript /tmp/tmpipvbivnyjlink/runner.jlink -nogui 1
```

The only thing I can notice from this output is that the commander script file is not generated, I might need to debug that python script, but other than that this doesn't tell me anything.

Regards.

**Reply by thedjnK** [reply, marked as answer]
> VTref=0.000V

No device is connected

**Reply by workholic365** [reply]
Hi @thedjnK  and @DerekSnell , @DerekSnell 
    I also met this issue when I execute west flash on imx93evk board .the detailed log as following with command :"west -v flash --runner jlink"

.venv) dell@dell-VirtualBox:~/zephyrproject/zephyr$ west -v flash --runner jlink
-- west flash: rebuilding
cmake version 3.22.1 is OK; minimum version is 3.13.1
Running CMake: /usr/bin/cmake --build /home/dell/zephyrproject/zephyr/build
ninja: no work to do.
-- west flash: using runner jlink
runners.jlink: JLink executable: /opt/SEGGER/JLink_V862/JLinkExe
runners.jlink: JLINKARM_GetDLLVersion()=86200
-- runners.jlink: JLink version: 8.62
runners.jlink: JLink commander script:
ExitOnError 1
r
LE
loadfile "/home/dell/zephyrproject/zephyr/build/zephyr/zephyr.bin" 0xd0000000
WReg PC 0xd0000000
g
writeDP 1 0
readDP 1
q

-- runners.jlink: Flashing file: /home/dell/zephyrproject/zephyr/build/zephyr/zephyr.bin

runners.jlink: /opt/SEGGER/JLink_V862/JLinkExe -nogui 1 -if swd -speed auto -device MIMX9352_A55_0 -CommanderScript /tmp/tmpsmpfy8i5jlink/runner.jlink -nogui 1
SEGGER J-Link Commander V8.62 (Compiled Aug 20 2025 11:14:38)
DLL version V8.62, compiled Aug 20 2025 11:13:35

J-Link Command File read successfully.
Processing script file...
J-Link>ExitOnError 1
J-Link Commander will now exit on Error
J-Link>r
J-Link connection not established yet but required for command.
Connecting to J-Link via USB...O.K.
Firmware: J-Link V13 compiled Jul 21 2025 13:16:03
Hardware version: V13.00
J-Link uptime (since boot): 0d 18h 04m 24s
S/N: 603002565
License(s): RDI, FlashBP, FlashDL, JFlash, GDB
USB speed mode: High speed (480 MBit/s)

**VTref=1.802V**

Target connection not established yet but required for command.

Device "MIMX9352_A55_0" selected.

Connecting to target via SWD
ConfigTargetSettings() start
ConfigTargetSettings() end - Took 18us
Found SW-DP with ID 0x5BA02477
DPIDR: 0x5BA02477
CoreSight SoC-400 or earlier
AP map detection skipped. Manually configured AP map found.
AP[0]: AHB-AP (IDR: Not set, ADDR: 0x00000000)
AP[1]: APB-AP (IDR: Not set, ADDR: 0x00000000)
Using preconfigured AP[1] as APB-AP
AP[1]: APB-AP found
DebugRegs + CTI manually specified. ROM table scan skipped.
Cortex-A55 @ 0x80810000 (configured)
CoreCTI @ 0x80820000 (configured)
Debug architecture: ARMv8.2
6 code breakpoints, 4 data breakpoints
Processor features:
  EL0 support: AArch64 + AArch32
  EL1 support: AArch64 + AArch32
  EL2 support: AArch64 + AArch32
  EL3 support: AArch64 + AArch32
  FPU support: Single + Double + Conversion + single arithmetic
Add. info (CPU temp. halted)
Current exception level: EL1
Exception level AArch usage:
  EL0: AArch64
  EL1: AArch64
  EL2: AArch64
  EL3: AArch64
Non-secure status: Non-secure
Cache info:
  Inner cache boundary: L3
  LoU Uniprocessor: 0
  LoC: 3
  LoU Inner Shareable: 0
I-Cache L1: 32 KB, 128 Sets, 64 Bytes/Line, 4-Way
D-Cache L1: 32 KB, 128 Sets, 64 Bytes/Line, 4-Way
Unified-Cache L2: 64 KB, 256 Sets, 64 Bytes/Line, 4-Way
Unified-Cache L3: 256 KB, 256 Sets, 64 Bytes/Line, 16-Way
VMSAv8-64: Supports 48-bit VAs
Memory zones:
  Zone: "Default" Description: Default access mode
  Zone: "AP0" Description: MEM-AP (AHB-AP)
  Zone: "AP1" Description: MEM-AP (APB-AP)
Cortex-A55 identified.
Reset delay: 0 ms
ResetTarget() start
No Reset strategy is implemented. Reset will be skipped.
ResetTarget() end - Took 25.0ms
Device specific reset executed.
J-Link>LE
J-Link>loadfile "/home/dell/zephyrproject/zephyr/build/zephyr/zephyr.bin" 0xd0000000
'loadfile': Performing implicit reset & halt of MCU.
ResetTarget() start
No Reset strategy is implemented. Reset will be skipped.
ResetTarget() end - Took 9us
Device specific reset executed.
Downloading file [/home/dell/zephyrproject/zephyr/build/zephyr/zephyr.bin]...
Writing target memory failed.
Script processing completed.

**FATAL ERROR**: command exited with status 1: /opt/SEGGER/JLink_V862/JLinkExe -nogui 1 -if swd -speed auto -device MIMX9352_A55_0 -CommanderScript /tmp/tmpsmpfy8i5jlink/runner.jlink -nogui 1

could you help check and give some suggestions ?
Many thinks !

**Reply by zczhou314** [reply]
For iMX93, "west flash" does not really burn to flash, it only loads the zephyr binary to DDR to run, as you can see from the log, it loads to DDR address 0xd0000000. Personally I always uses tftp under u-boot to run zephyr.  If you still want to use JLink to load Zephyr, I can try to dig in. Please let me know.

**Reply by workholic365** [reply]
Hi @zczhou314 
    Thanks for your reply . 
    Currently ,I run the zephyr.bin with TF card .  first to copy the zephyr.bin to TF card .and then power on ,after that interrupt in boot mode .then execute the command as following :
    `fatload mmc 1:1 0xd0000000 zephyr.bin; dcache off; icache flush; go 0xd0000000;`
    maybe my current operation like yours . just use TF card insteads of tftp . 
    I hope could use west flash to realize the same function . 
    I will appreciate your help and try .

**Comment by emmdaniel** [comment]
@thedjnK That was the issue, thanks!
